### PR TITLE
READY: Change Matplotlib graph to web-based setup using QWebEngineView.

### PR DESCRIPTION
**Pull request by rkeulemans**
Solves #77

### CODE REVIEW

File: Tribler/Core/Modules/restapi/display_endpoint.py [hunk 1]
```diff
@@ -107,6 +107,9 @@ def render_GET(self, request):
         :param request: the HTTP GET request which specifies the focus node and optionally the neighbor level
         :return: the node data formatted in JSON
         """
+
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/display_endpoint.py]: Please remove the above newline and add a comment explaining why this header is necessary.

File: Tribler/Core/Modules/restapi/display_endpoint.py [hunk 2]
```diff
@@ -126,14 +129,18 @@ def render_GET(self, request):
         else:
             return DisplayEndpoint.return_error(request, message="focus_node was not a string")
 
+        print "--------------", focus_node, type(focus_node)
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/display_endpoint.py]: What is this print statement?

File: Tribler/Core/Modules/restapi/display_endpoint.py [hunk 3]
```diff
@@ -126,14 +129,18 @@ def render_GET(self, request):
...
+
         neighbor_level = 1
         # Note that isdigit() checks if all chars are numbers, hence negative numbers are not possible to be set
         if "neighbor_level" in request.args and len(request.args["neighbor_level"]) > 0 and \
                 request.args["neighbor_level"][0].isdigit():
             neighbor_level = int(request.args["neighbor_level"][0])
 
-        # TODO: Remove dummy return and change to aggregated data calculation
+        mc_community = self.get_multi_chain_community()
+        (nodes, edges) = mc_community.get_graph(focus_node, neighbor_level)
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/display_endpoint.py]: Please remove these parenthesis.

File: TriblerGUI/tribler_app.py [hunk 4]
```diff
@@ -2,6 +2,7 @@
 
 from PyQt5.QtCore import QEvent
 from TriblerGUI.single_application import QtSingleApplication
+from PyQt5.QtWebEngineWidgets import QWebEngineView
```
Comment (vandenheuvel) [inline TriblerGUI/tribler_app.py]: Can you place a comment explaining why this import is necessary?

File: TriblerGUI/widgets/trustpage/graph.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+{
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/graph.json]: Why is this fake data in a `READY` PR?

File: TriblerGUI/widgets/trustpage/index.html [hunk 6]
```diff
@@ -0,0 +1,19 @@
+<!DOCTYPE html>
+<html lang="en">
+<head>
+    <meta charset="UTF-8">
+    <title>Trust Page</title>
+
+    <!-- Import local copy of D3.js to speed up loading -->
+    <!-- TODO Change to CDN or NPM dependency -->
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/index.html]: I don't think we want to use NPM packages as NPM won't be shipped with Tribler.

File: TriblerGUI/widgets/trustpage/js/request_data.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+function make_cors_request(method, url) {
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/request_data.js]: Please add a docstring explaining what is in this file.

File: TriblerGUI/widgets/trustpage/js/main.js [hunk 8]
```diff
@@ -2,6 +2,10 @@
  * This file creates an SVG element and draws the network using it.
  */
 
+// TODO: think about how to split this in different javascipt files
+// TODO: start using JSDoc and setup some CI tool to check
+// TODO: choose a testing framework and start writing tests
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/main.js]: I don't think these `TODO`'s belong in the codebase. I suggest you create tickets for these.

File: TriblerGUI/widgets/trustpage/js/main.js [hunk 9]
```diff
@@ -20,10 +24,11 @@ var vis = d3.select("#graph")
...
 
-// When JSON is read, update the display
-d3.json("graph.json", function (data) {
-    update(data.nodes);
-});
+function new_data(nodes) {
+    update(JSON.parse(nodes).nodes)
+}
+
+get_node_info(30, new_data);
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/main.js]: What is this magic 30?

File: TriblerGUI/widgets/trustpage/js/request_data.js [hunk 10]
```diff
@@ -20,7 +25,7 @@ function get_node_info(public_key) {
     }
 
     response.onload = function() {
-        console.log(response.responseText);
+        callback(response.responseText);
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/request_data.js]: Instead of creating a function which calls the passed callback, can't you assign the callback directly to `response.onload`, while passing it's argument(s)?

File: Tribler/Core/Modules/restapi/display_endpoint.py [hunk 11]
```diff
@@ -89,7 +89,7 @@ def render_GET(self, request):
 
                 {
                     "focus_node": "xyz",
-                    "neighbor_level": 1,
+                    "neighbor_level: 1
                     "nodes": [{
```
Comment (yinghaodai) [inline Tribler/Core/Modules/restapi/display_endpoint.py]: Please reset this

**Comment by vandenheuvel** [comment]
@TimSpeelman Please fix the tests.

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+    // Create an <svg.node> element.
+    var groups = selection
+        .append("svg")
+        .attr("overflow", "visible")
+        .attr("class", "node");
+
+    // Append a <circle> element to it.
+    var circles = groups
+        .append("circle")
+        .attr("fill", "rgb(230,115,0)")
```
Comment (ghabbenjansen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Why are you using RGB here but hex on line 55?

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+    var selection = selectLinks(svg).data(data).enter();
+
+    var links = selection
+        .append("svg")
+        .attr("class", "link");
+
+    links.append("line")
+        .attr("class", "link-source")
+        .attr("stroke-width", 2)
+        .style("stroke", "yellow");
```
Comment (ghabbenjansen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: And why are you using a colour name here and hex earlier

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+
+    links.append("line")
+        .attr("class", "link-target")
+        .attr("stroke-width", 2)
+        .style("stroke", "red");
```
Comment (ghabbenjansen) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: See above

File: TriblerGUI/widgets/trustpage/js/main.js [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+    });
+
+    // All nodes start in the center (slightly off)
+    graph.nodes.forEach(function (node, i) {
+        node.x = width / 2 + Math.random();
+        node.y = height / 2 + Math.random();
+    });
+
+    // Position all direct neighbors on a circle
+    const pi = Math.PI;
```
Comment (ghabbenjansen) [inline TriblerGUI/widgets/trustpage/js/main.js]: Can't you use Math.PI in the code instead of pi

Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/main.js]: I think pi is more readable.

File: Tribler/Core/Modules/restapi/display_endpoint.py [hunk 16]
```diff
@@ -89,7 +89,7 @@ def render_GET(self, request):
 
                 {
                     "focus_node": "xyz",
-                    "neighbor_level": 1,
+                    "neighbor_level": 1
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/display_endpoint.py]: Please revert this change.

File: Tribler/Core/Modules/restapi/display_endpoint.py [hunk 17]
```diff
@@ -133,7 +136,9 @@ def render_GET(self, request):
             neighbor_level = int(request.args["neighbor_level"][0])
 
         mc_community = self.get_multi_chain_community()
+
```
Comment (vandenheuvel) [inline Tribler/Core/Modules/restapi/display_endpoint.py]: Please remove this white space.

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 18]
```diff
@@ -0,0 +1,104 @@
+/**
+ * These methods are concerned with drawing and styling SVG elements
+ */
+
+/**
+ * Select all <svg.node> elements in .nodes
+ * @returns {*}
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Please add a docstring for the parameter.

Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Could you add a newline before the `@param` and `@returns` lines?

File: TriblerGUI/widgets/trustpage/js/drawing.js [hunk 19]
```diff
@@ -0,0 +1,104 @@
...
+
+    // Append a <text> element to it
+    groups
+        .append("text")
+        .attr("x", 24)
+        .attr("y", 24)
+        .style("font-family", "sans-serif")
+        .style("font-size", "12")
+        .style("fill", "#ffff00")
+        .text(function (d) {
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: Are there java8-like shortcuts for these like `d -> d.public-key`?

File: TriblerGUI/widgets/trustpage/js/main.js [hunk 20]
```diff
@@ -0,0 +1,189 @@
+/**
+ * This file creates an SVG element and draws the network using it.
+ *
+ * NOTE: Use a browser DOM inspection tool to see how the SVG is
+ * built up and modified by this code.
+ */
+
+// TODO: start using JSDoc and setup some CI tool to check
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/main.js]: You could add documentation for these functions already, right?

File: TriblerGUI/widgets/trustpage/js/main.js [hunk 21]
```diff
@@ -0,0 +1,189 @@
...
+    .force("center", d3.forceCenter(state.x, state.y))
+
+    // Centering the neighbor nodes (radial layout)
+    .force("neighbor_x", d3.forceX(getRadialPosition(0)).strength(.5))
+    .force("neighbor_y", d3.forceY(getRadialPosition(1)).strength(.5))
+
+    // The update function for every tick of the clock
+    .on("tick", tick)
+
+    .alphaDecay(0);
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/main.js]: Could you add another single-line doc string?

File: TriblerGUI/widgets/trustpage/js/process_data.js [hunk 22]
```diff
@@ -0,0 +1,130 @@
...
+    var data = JSON.parse(jsonData);
+
+    // Map edges to links
+    data.links = data.edges;
+
+    var groupedEdges = groupBy(data.links, "source");
+    var combinedEdges = [];
+
+
+
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/process_data.js]: Too many newlines.

File: TriblerGUI/widgets/trustpage/js/radial.js [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+ * @param center_x
+ * @param center_y
+ * @param alpha
+ * @param radius
+ * @returns {{x: *, y: number}}
+ */
+function polarToCartesian(center_x, center_y, alpha, radius) {
+    return {
+        x: center_x + radius * Math.cos(alpha),
+        y: center_y - radius * Math.sin(alpha) // Minus because screen Y points down
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/radial.js]: Could you move this comment above this line?

File: TriblerGUI/widgets/trustpage/js/utilities.js [hunk 24]
```diff
@@ -0,0 +1,48 @@
+/**
```
Comment (vandenheuvel) [inline TriblerGUI/widgets/trustpage/js/utilities.js]: Please add a docstring explaining what is in this file.

File: TriblerGUI/widgets/trustpage/js/drawing.js (hunk 19 above)
Comment (TimSpeelman) [inline TriblerGUI/widgets/trustpage/js/drawing.js]: In ES6 only

**Comment by MaxVanDeursen** [comment]
The GUI tests can be fixed by adding the QtWebEngineWidgets import to test_gui.py, for the same reason as given in the tribler_app.py. 

Moreover, I think you should be able to boost that coverage difference.

**Comment by MaxVanDeursen** [comment]
Note about PyLint error: 

Although the imported module is not used in this class, it is necessary to import the QtWebEngineWidgets here because otherwise an ImportError will be thrown. This is troubleshooted extensively in [this StackOverflow post](https://stackoverflow.com/questions/41172301/pyqt-modules-cant-be-imported-after-qtwidgets-qapplicationsys-argv) and seems to be necessary.
